### ISSUE TITLE
Installing/using in internal network only

### ISSUE THREAD

**Issue by notDavid** [searx/searx-docker#22]
Hi there, i have a question and i'm hoping you could point me in the right direction;

I'd like to install searx-docker in an internal network, where port 80 and 443 are not accessible from the internet (which is required for letsencrypt to generate (and renew?) the certificate.)

I could open the ports only while installing, to generate the certificate, but then:
- i would need the certificate to be valid for a long time like a few years so it doesn't need to renew
- i assume the certificate won't work anyway on an internal net since the domain name will then point to a (no longer public) different internal ip address?

Long story short, what config/settings are best for use in an internal network only? Can i use http only?

Thank you.

**Comment by dalf** [comment]
I've never done that, but be prepare to have different result according to the browser.

You may need to provide [OCSP stapling](https://en.wikipedia.org/wiki/OCSP_stapling) on the server.
Otherwise [OCSP](https://en.wikipedia.org/wiki/Online_Certificate_Status_Protocol) (on the browser) may cause some issue if the client has no Internet access.

>  Can i use http only?

You can use http only if you trust all the computers connected to internal network you are describing.
Trust = the computers won't spy / MITM your connections.

> i assume the certificate won't work anyway on an internal net since the domain name will then point to a (no longer public) different internal ip address?

If you have a public DNS resolver and internal DNS resolver you can do that:
* return a public IP to resolve the HTTP / DNS challenge for letsencrypt, so your server has the certificate and the private key. The public IP is only here to solve the challenge, no other public service.
* return the private IP in your network.

> i would need the certificate to be valid for a long time like a few years so it doesn't need to renew

Then let's encrypt won't work for you. 
You can buy / get a certificate from a different cert authority: you will get a private key and the certificate manually so you can install it.

See: https://security.stackexchange.com/questions/121163/how-do-i-run-proper-https-on-an-internal-network

**Comment by notDavid** [comment]
@dalf Thank you for the info! I think http is not an option after all, because many browsers are starting to complain about http urls these days; So i'm going to try to create a certificate for the internal domain.

My clients have full internet access. The only problem is that the firewall can not forward ports 80 and 443 to the Searx server (which is required for Let's encrypt.) So those ports aren't accessible from the outside. I can forward them temporarily during install, but not permanently.
So i probably need to buy a real certificate then...

**Comment by notDavid** [comment]
@dalf fyi i found out that Caddy supports the letsencrypt "DNS Challenge":

> Caddy solves the DNS challenge which **does not involve opening any ports** on the machine. There are integrations for all major DNS providers!

https://github.com/caddyserver/dnsproviders/blob/master/README.md

I'm not sure if i can figure out how to implement it, will have a look at it later.
